drivers/net/mlx5/mlx5_vdpa.c

```diff
@@ -147,14 +141,81 @@ mlx5_vdpa_query_virtio_caps(struct vdpa_priv *priv)
 	 * For now only set protocol features support
 	 */
 	priv->caps.max_num_virtqs = MLX5_VDPA_SW_MAX_VIRTQS_SUPPORTED;
-	priv->caps.virtio_net_features = (1ULL << VHOST_USER_F_PROTOCOL_FEATURES);
+	priv->caps.virtio_net_features = MLX5_VDPA_FEATURES;
 	DRV_LOG(DEBUG, "Virtio Caps:");
 	DRV_LOG(DEBUG, "	dump_mkey=0x%x ", priv->caps.dump_mkey);
 	DRV_LOG(DEBUG, "	max_num_virtqs=0x%x ", priv->caps.max_num_virtqs);
 	DRV_LOG(DEBUG, "	features_bits=0x%" PRIx64, priv->caps.virtio_net_features);
 	return 0;
 }
 
+#define MLX5_IB_MMAP_CMD_SHIFT 8
+#define MLX5_IB_MMAP_CMD_MASK ((1 << MLX5_IB_MMAP_CMD_SHIFT) - 1)
+#define MLX5_IB_CMD_SIZE 8
+#define MLX5_IB_MMAP_VIRTIO_NOTIFY 9
+static inline int
+mlx5_vdpa_get_notify_offset(uint16_t offset_cap)
+{
+	uint16_t ext_offset = MLX5_IB_MMAP_CMD_SHIFT + MLX5_IB_CMD_SIZE;
+	return (((offset_cap >> MLX5_IB_MMAP_CMD_SHIFT) << ext_offset) |
+		(MLX5_IB_MMAP_VIRTIO_NOTIFY << MLX5_IB_MMAP_CMD_SHIFT) |
+		(offset_cap & MLX5_IB_MMAP_CMD_MASK));
+}
```

Comment:
We need to take the queue index here to pick the right stride inside the bar.
Also I don't understand how this can work. the cmd_fd is a UAR page, and offset cap you get from FW
is absolute offset from start of BAR0.
So we need to assume we have a new fd which is the correct page from the BAR, and then you need to set the offset from within the page according to the queue index. 
Right?

Reply:
This is not correct. the cmd_fd maps the device bar starting from the initialization segment.